[question] How to deal with "faux" transitive dependency from wrapper package

### What is your question?

Hi,

I have this wrapper package for Qt installed locally via Qt Online Installer:
<details> <summary> qt wrapper recipe </summary> 

```
from conan import ConanFile
from conan.errors import ConanInvalidConfiguration
from conan.tools.files import collect_libs
import os
import platform

class QtWrapperConan(ConanFile):
    name = "qt"
    license = "Commercial"
    description = "Wrapper for Qt installed via Qt Online Installer"
    settings = "os", "compiler", "arch"
    options = {
        "qt_prefix": ["ANY", None],
    }
    default_options = {
        "qt_prefix": None,
    }
    package_type = "shared-library"
    build_policy = "never"

    def _get_qt_suffix(self):
        # Mapping (compiler, version) -> Qt suffix
        msvc_map = {
            "msvc": {
                "192": "msvc2019",
                "193": "msvc2022",
            },
            "Visual Studio": {
                "16": "msvc2019",
                "17": "msvc2022",
            },
        }

        compiler = str(self.settings.compiler)
        version = str(self.settings.compiler.version)

        try:
            return msvc_map[compiler][version]
        except KeyError:
            raise ConanInvalidConfiguration(f"Unsupported compiler/version combo: {compiler} {version}")

    def layout(self):
        pass  # Not needed

    def validate(self):
        if self.options.qt_prefix:
            if not os.path.isdir(str(self.options.qt_prefix)):
                raise ConanInvalidConfiguration(f"qt_prefix does not exist: {self.options.qt_prefix}")

    def package_info(self):
        self.cpp_info.set_property("cmake_find_mode", "none")

        if self.options.qt_prefix:
            qt_path = os.path.abspath(str(self.options.qt_prefix))
        else:
            if platform.system() == "Windows":
                suffix = self._get_qt_suffix()
                qt_path = os.path.expanduser(os.path.join("C:/Qt", self.version, suffix))
            else:
                qt_path = os.path.expanduser(os.path.join("~/Qt", self.version, "gcc_64"))

        self.output.info(f"Using Qt path: {qt_path}")
        self.cpp_info.builddirs = [os.path.join(qt_path, "lib", "cmake"), os.path.join(qt_path, "lib", "pkgconfig")]

```

</details>


The main point is that it just uses cmake/pkgconfig files already preinstalled:
```
        self.cpp_info.set_property("cmake_find_mode", "none")
```
```
        self.cpp_info.builddirs = [os.path.join(qt_path, "lib", "cmake"), os.path.join(qt_path, "lib", "pkgconfig")]
```

Now, my KArchive uses that qt dependency:

<details> <summary> KArchive recipe </summary>

```
import os
import pathlib
from conan import ConanFile
from conan.tools.files import get, copy, rmdir, replace_in_file
from conan.tools.cmake import CMakeToolchain, CMakeDeps, CMake, cmake_layout
from conan.errors import ConanInvalidConfiguration


class KArchiveConan(ConanFile):
    name = "karchive"
    version = "5.77.0"
    package_type = "library"
    license = "LGPL-2.1"
    url = "https://api.kde.org/frameworks/karchive/html/index.html"
    description = "Qt-based library for reading/writing archive formats"
    settings = "os", "compiler", "build_type", "arch"
    options = {
        "shared": [True, False],
        "with_bzip2": [True, False],
        "with_lzma": [True, False],
    }
    default_options = {
        "shared": True,
        "with_bzip2": False,
        "with_lzma": False
    }

    def layout(self):
        cmake_layout(self)

    def build_requirements(self):
        self.tool_requires(f"extra-cmake-modules/{self.version}@mycompany/stable")

    def requirements(self):
        self.requires("qt/[>=5 <6]@qtgroup/installed")
        self.requires("zlib/1.2.11")
        if self.options.with_bzip2:
            self.requires("bzip2/1.0.8")
        if self.options.with_lzma:
            self.requires("xz_utils/5.2.5")  # note: modern name for lzma

    def source(self):
        get(self,
            url=f"https://github.com/KDE/karchive/archive/v{self.version}.tar.gz",
            sha256="ff0047cd0adab5e451b98ac2ea73339b38a048e1ccc85666f8e81952ce147d18",
            strip_root=True)
        
        # make it work with Conan dependencies
        replace_in_file(self, os.path.join(self.source_folder, "src", "CMakeLists.txt"),
                        "${ZLIB_LIBRARY}",
                        "ZLIB::ZLIB")

    def generate(self):
        tc = CMakeToolchain(self)
        tc.variables["BUILD_QCH"] = False
        tc.variables["BUILD_TESTING"] = False
        tc.variables["ECM_DIR"] = pathlib.Path(
            os.path.join(self.dependencies.build["extra-cmake-modules"].package_folder, "res", "ECM", "cmake")
        ).as_posix()
        
        # Disable system optional packages when option is off
        if not self.options.with_bzip2:
            tc.variables["CMAKE_DISABLE_FIND_PACKAGE_BZip2"] = True
        if not self.options.with_lzma:
            tc.variables["CMAKE_DISABLE_FIND_PACKAGE_LibLZMA"] = True
        
        tc.generate()

        deps = CMakeDeps(self)
        deps.generate()

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()
        cmake.install()

    def package(self):
        copy(self, "*", dst=os.path.join(self.package_folder, "licenses"), src=os.path.join(self.source_folder, "LICENSES"))
        rmdir(self, os.path.join(self.package_folder, "lib", "cmake"))
        rmdir(self, os.path.join(self.package_folder, "mkspecs"))

    def package_info(self):
        self.cpp_info.libs = ["KF5Archive"]
        self.cpp_info.includedirs = ["include/KF5", "include/KF5/KArchive"]
        self.cpp_info.set_property("cmake_file_name", "KArchive")
        self.cpp_info.set_property("cmake_target_name", "KArchive::KArchive")
```

</details>

Key point that it requries that qt wrapper:

```
 self.requires("qt/[>=5 <6]@qtgroup/installed")
```

Now, if my application also requires qt:

```
from conan import ConanFile
from conan.tools.files import copy
import os

DEFAULT_QT_VERSION = "5.15.19"

class TestBackupConan(ConanFile):
    name = "TestBackup"
    settings = "os", "compiler", "arch", "build_type"
    generators = "CMakeToolchain", "CMakeDeps"

    def requirements(self):
        qt_version = os.getenv("STY_OVERRIDE_QT_VERSION", DEFAULT_QT_VERSION)
        self.requires(f"qt/{qt_version}@qtgroup/installed")
        self.requires("karchive/5.77.0@mycompany/stable")

    def generate(self):
        dest_folder = os.path.join(self.build_folder, '..')
        
        for dep in self.dependencies.values():
            copy(self, "KF5Archive*.dll", src=dep.cpp_info.bindirs[0], dst=dest_folder)
            copy(self, "zlib*.dll", src=dep.cpp_info.bindirs[0], dst=dest_folder)
            
            copy(self, "libKF5Archive.so*", src=dep.cpp_info.libdirs[0], dst=dest_folder)
            copy(self, "libz.so*", src=dep.cpp_info.libdirs[0], dst=dest_folder)

```

When CMake is run, I get strange "requirement" `qt`:

```
[cmake] -- Using Conan toolchain: /home/path/_conan/conan_toolchain.cmake
[cmake] -- Conan toolchain: Defining architecture flag: -m64
[cmake] -- Conan toolchain: C++ Standard 20 with extensions ON
[cmake] -- Conan: Target declared 'KArchive::KArchive'
[cmake] CMake Error at /usr/share/cmake-3.25/Modules/CMakeFindDependencyMacro.cmake:47 (find_package):
[cmake]   By not providing "Findqt.cmake" in CMAKE_MODULE_PATH this project has asked
[cmake]   CMake to find a package configuration file provided by "qt", but CMake did
[cmake]   not find one.
[cmake] 
[cmake]   Could not find a package configuration file provided by "qt" with any of
[cmake]   the following names:
[cmake] 
[cmake]     qtConfig.cmake
[cmake]     qt-config.cmake
[cmake] 
[cmake]   Add the installation prefix of "qt" to CMAKE_PREFIX_PATH or set "qt_DIR" to
[cmake]   a directory containing one of the above files.  If "qt" provides a separate
[cmake]   development package or SDK, be sure it has been installed.
[cmake] Call Stack (most recent call first):
[cmake]   /home/path/_conan/KArchiveConfig.cmake:24 (find_dependency)
[cmake]   /home/other_path/Backup/CMakeLists.txt:14 (find_package)
[cmake] 
[cmake] 
[cmake] -- Configuring incomplete, errors occurred!
```

My workaround is to set `visible=False` in KArchive recipe:
```
self.requires("qt/[>=5 <6]@qtgroup/installed", visible=False)
```
Basically, there's no problem if only the application uses Qt wrapper package, but **once at least one Conan-dependency ALSO uses same Qt wrapper pacakge, I get this strange issue.**

Note that there's no `qtConfig.cmake` in preinstalled Qt directory, so it's not "expected" in any way. There's `/home/vincas/Qt/5.15.19/gcc_64/lib/cmake/Qt5/Qt5Config.cmake` and other `Qt5*`.

</details>

While `visible=False` works, I doubt if it's right thing do.

Imagine as if `KArchvie` would be upstreamed into `conan-center-index`, so it should work with "normal" Qt package, AND with my local override (via `[replace_requires]`).

So where's the problem? I hope it's not a Conan bug..?

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

I am trying to reproduce, but in the meantime I have seen:

> CMake Error at /usr/share/cmake-3.25/Modules/CMakeFindDependencyMacro.cmake:47 (find_package):

This is usually an indicator that:

- The consumer ``CMakeLists.txt`` is doing a bare ``find_package(qt)``, without explicitly doing ``find_package(qt CONFIG REQUIRED)``
- CMake will prioritize its own internal FindQt.cmake module, not the ``qtConfig.cmake`` one inside the Conan package

Could you please share your consuming ``CMakeLists.txt`` including the ``find_package()`` of the KArchive and Qt?

--- a/test/functional/toolchains/cmake/test_cmake_find_none.py
+++ b/test/functional/toolchains/cmake/test_cmake_find_none.py
@@ -3,7 +3,7 @@ import textwrap
 from conan.test.utils.tools import TestClient
 
 
-def test_cmake_find_none():
+def test_cmake_find_none_transitive():
     c = TestClient()
 
     qt = textwrap.dedent("""
@@ -100,5 +100,5 @@ def test_cmake_find_none():
     c.save({"conanfile.py": consumer,
             "CMakeLists.txt": consumer_cmake}, clean_first=True)
     c.run("build .")
-    print(c.out)
-
+    assert "Conan: Target declared 'karchive::karchive'" in c.out
+    # And it doesn't fail to find transitive qt

--- a/test/functional/toolchains/cmake/test_cmake_find_none.py
+++ b/test/functional/toolchains/cmake/test_cmake_find_none.py
@@ -34,6 +34,7 @@ def test_cmake_find_none():
             def package_info(self):
                 self.cpp_info.builddirs = ["qt/cmake"]
                 self.cpp_info.set_property("cmake_find_mode", "none")
+                self.cpp_info.set_property("cmake_file_name", "Qt5")
         """)
 
     cmake = textwrap.dedent("""
@@ -74,7 +75,7 @@ def test_cmake_find_none():
         cmake_minimum_required(VERSION 3.15)
         project(MyHello NONE)
 
-        find_package(Qt5 CONFIG REQUIRED)
+        find_package(karchive CONFIG REQUIRED)
         """)
 
     consumer = textwrap.dedent("""
@@ -84,11 +85,11 @@ def test_cmake_find_none():
         class pkgRecipe(ConanFile):
             package_type = "static-library"
             settings = "os", "compiler", "build_type", "arch"
-            generators = "CMakeToolchain"
+            generators = "CMakeToolchain", "CMakeDeps"
 
             def requirements(self):
-                self.requires("qt/0.1")
                 self.requires("karchive/0.1")
+                self.requires("qt/0.1")
 
             def build(self):
                 cmake = CMake(self)

--- a/test/functional/toolchains/cmake/test_cmake_find_none.py
+++ b/test/functional/toolchains/cmake/test_cmake_find_none.py
@@ -1,0 +1,103 @@
+import textwrap
+
+from conan.test.utils.tools import TestClient
+
+
+def test_cmake_find_none():
+    c = TestClient()
+
+    qt = textwrap.dedent("""
+        from conan import ConanFile
+        from conan.tools.cmake import CMake
+
+        class pkgRecipe(ConanFile):
+            name = "qt"
+            version = "0.1"
+            package_type = "static-library"
+
+            # Binary configuration
+            settings = "os", "compiler", "build_type", "arch"
+            generators = "CMakeToolchain"
+
+            # Sources are located in the same place as this recipe, copy them to the recipe
+            exports_sources = "CMakeLists.txt", "src/*", "include/*"
+
+            def build(self):
+                cmake = CMake(self)
+                cmake.configure()
+                cmake.build()
+
+            def package(self):
+                cmake = CMake(self)
+                cmake.install()
+
+            def package_info(self):
+                self.cpp_info.builddirs = ["qt/cmake"]
+                self.cpp_info.set_property("cmake_find_mode", "none")
+        """)
+
+    cmake = textwrap.dedent("""
+        cmake_minimum_required(VERSION 3.15)
+        project(MyHello NONE)
+
+        add_library(qt INTERFACE)
+        install(TARGETS qt EXPORT qtConfig)
+        export(TARGETS qt
+            NAMESPACE qt::
+            FILE "${CMAKE_CURRENT_BINARY_DIR}/qtConfig.cmake"
+        )
+        install(EXPORT qtConfig
+            DESTINATION "${CMAKE_INSTALL_PREFIX}/qt/cmake"
+            NAMESPACE qt::
+        )
+        """)
+    c.save({"conanfile.py": qt,
+            "CMakeLists.txt": cmake})
+    c.run("create .")
+
+    karchive = textwrap.dedent("""
+        from conan import ConanFile
+        from conan.tools.cmake import CMake
+
+        class pkgRecipe(ConanFile):
+            name = "karchive"
+            version = "0.1"
+            package_type = "static-library"
+
+            requires = "qt/0.1"
+        """)
+
+    c.save({"conanfile.py": karchive}, clean_first=True)
+    c.run("create .")
+
+    consumer_cmake = textwrap.dedent("""
+        cmake_minimum_required(VERSION 3.15)
+        project(MyHello NONE)
+
+        find_package(qt CONFIG REQUIRED)
+        """)
+
+    consumer = textwrap.dedent("""
+        from conan import ConanFile
+        from conan.tools.cmake import CMake
+
+        class pkgRecipe(ConanFile):
+            package_type = "static-library"
+            settings = "os", "compiler", "build_type", "arch"
+            generators = "CMakeToolchain"
+
+            def requirements(self):
+                self.requires("qt/0.1")
+                self.requires("karchive/0.1")
+
+            def build(self):
+                cmake = CMake(self)
+                cmake.configure()
+                cmake.build()
+        """)
+
+    c.save({"conanfile.py": consumer,
+            "CMakeLists.txt": consumer_cmake}, clean_first=True)
+    c.run("build .")
+    print(c.out)
+

--- a/test/functional/toolchains/cmake/test_cmake_find_none.py
+++ b/test/functional/toolchains/cmake/test_cmake_find_none.py
@@ -41,12 +41,12 @@ def test_cmake_find_none():
         project(MyHello NONE)
 
         add_library(qt INTERFACE)
-        install(TARGETS qt EXPORT qtConfig)
+        install(TARGETS qt EXPORT Qt5Config)
         export(TARGETS qt
             NAMESPACE qt::
-            FILE "${CMAKE_CURRENT_BINARY_DIR}/qtConfig.cmake"
+            FILE "${CMAKE_CURRENT_BINARY_DIR}/Qt5Config.cmake"
         )
-        install(EXPORT qtConfig
+        install(EXPORT Qt5Config
             DESTINATION "${CMAKE_INSTALL_PREFIX}/qt/cmake"
             NAMESPACE qt::
         )
@@ -74,7 +74,7 @@ def test_cmake_find_none():
         cmake_minimum_required(VERSION 3.15)
         project(MyHello NONE)
 
-        find_package(qt CONFIG REQUIRED)
+        find_package(Qt5 CONFIG REQUIRED)
         """)
 
     consumer = textwrap.dedent("""

--- a/test/functional/toolchains/cmake/test_cmake_toolchain.py
+++ b/test/functional/toolchains/cmake/test_cmake_toolchain.py
@@ -2063,3 +2063,37 @@ def test_cmake_toolchain_language_c():
     else:
         client.run("build . -s compiler.cppstd=11")
         # It doesn't fail
+
+
+@pytest.mark.skipif(platform.system() != "Windows", reason="Only for windows")
+@pytest.mark.tool("cmake")
+def test_cmaketoolchain_check_function_exists():
+    c = TestClient()
+    conanfile = textwrap.dedent("""
+        from conan import ConanFile
+        from conan.tools.cmake import CMake, CMakeToolchain
+        class Conan(ConanFile):
+            settings = "os", "compiler", "arch", "build_type"
+            def generate(self):
+                tc = CMakeToolchain(self)
+                # This gets back to older default behavior
+                # tc.blocks["try_compile"].values["config"] = None
+                tc.generate()
+            def build(self):
+                CMake(self).configure()
+        """)
+    consumer = textwrap.dedent("""
+        cmake_minimum_required(VERSION 3.15)
+        project(MyHello C)
+        include(CheckFunctionExists)
+        check_function_exists(pow HAVE_POW)
+        if(NOT HAVE_POW)
+            message(FATAL_ERROR "Not math!!!!!")
+        endif()
+       """)
+    c.save({"conanfile.py": conanfile,
+            "CMakeLists.txt": consumer})
+
+    c.run("build . ")  # Release breaks the check
+    assert "Looking for pow - found" in c.out
+    # And it no longer crashes

--- a/test/functional/toolchains/cmake/test_cmake_toolchain.py
+++ b/test/functional/toolchains/cmake/test_cmake_toolchain.py
@@ -2068,6 +2068,15 @@ def test_cmake_toolchain_language_c():
 @pytest.mark.skipif(platform.system() != "Windows", reason="Only for windows")
 @pytest.mark.tool("cmake")
 def test_cmaketoolchain_check_function_exists():
+    """
+    covers the problematic check_function_exists() CMake DISCOURAGED check, that can fail
+    for Release configuration, due to CMake issues. This was broken when we tried to
+    add the automatic CMAKE_TRY_COMPILE_CONFIGURATION in release cases, as CheckFunctionExists.c
+    is optimized away or fails to compile in Release, but works in Debug.
+    This test is here to capture this legacy and problematic use case NOT to recommended this
+    as a good practice, rather then opposite, DONT DO IT
+    """
+
     c = TestClient()
     conanfile = textwrap.dedent("""
         from conan import ConanFile

